codec-http2/src/test/java/io/netty/handler/codec/http2/Http2MultiplexCodecTest.java

```diff
@@ -12,193 +12,208 @@
-        channel.pipeline().addLast(serverCodec);
-        http2HandlerCtx = channel.pipeline().context(serverCodec.connectionHandler());
-
-        // Handshake
-        verify(frameWriter).writeSettings(eq(http2HandlerCtx),
-            anyHttp2Settings(), anyChannelPromise());
-        verifyNoMoreInteractions(frameWriter);
-        channel.writeInbound(Http2CodecUtil.connectionPrefaceBuf());
-        frameListener.onSettingsRead(http2HandlerCtx, new Http2Settings());
-        verify(frameWriter).writeSettingsAck(eq(http2HandlerCtx), anyChannelPromise());
-        frameListener.onSettingsAckRead(http2HandlerCtx);
+    public void setUp() {
+        childChannelInitializer = new TestChannelInitializer();
+        parentChannel = new EmbeddedChannel();
+        parentChannel.connect(new InetSocketAddress(0));
+        parentChannel.pipeline().addLast(new Http2MultiplexCodec(true, null, childChannelInitializer));
     }
 
-    @After
-    public void tearDown() {
-        Object o;
-        while ((o = channel.readOutbound()) != null) {
-            ReferenceCountUtil.release(o);
-        }
-    }
+    // TODO(buchgr): Thread model of child channel
+    // TODO(buchgr): Flush from child channel
+    // TODO(buchgr): ChildChannel.childReadComplete()
+    // TODO(buchgr): GOAWAY Logic
+    // TODO(buchgr): Reset frame on close
+    // TODO(buchgr): Test ChannelConfig.setMaxMessagesPerRead
 
     @Test
-    public void startStop() throws Exception {
-        assertTrue(channel.isActive());
-        channel.close();
-        verify(frameWriter).writeGoAway(
-                eq(http2HandlerCtx), eq(0), eq(0L), eq(Unpooled.EMPTY_BUFFER), anyChannelPromise());
-        assertTrue(!channel.isActive());
+    public void headerAndDataFramesShouldBeDelivered() {
+        LastInboundHandler inboundHandler = new LastInboundHandler();
+        childChannelInitializer.handler = inboundHandler;
+
+        Http2StreamActiveEvent streamActive = new Http2StreamActiveEvent(streamId);
+        Http2HeadersFrame headersFrame = new DefaultHttp2HeadersFrame(request).setStreamId(streamId);
+        Http2DataFrame dataFrame1 = new DefaultHttp2DataFrame(bb("hello")).setStreamId(streamId);
+        Http2DataFrame dataFrame2 = new DefaultHttp2DataFrame(bb("world")).setStreamId(streamId);
+
+        assertFalse(inboundHandler.channelActive);
+        parentChannel.pipeline().fireUserEventTriggered(streamActive);
+        assertTrue(inboundHandler.channelActive);
+        // Make sure the stream active event is not delivered as a user event on the child channel.
+        assertNull(inboundHandler.readUserEvent());
+        parentChannel.pipeline().fireChannelRead(headersFrame);
+        parentChannel.pipeline().fireChannelRead(dataFrame1);
+        parentChannel.pipeline().fireChannelRead(dataFrame2);
+
+        assertEquals(headersFrame, inboundHandler.readInbound());
+        assertEquals(dataFrame1, inboundHandler.readInbound());
```

Comment:
IMO we should still release. test code should not be allocator specific if it doesn't need to be.

@normanmaurer - WDYT?